clusterloader2/pkg/config/codec.go

```diff
@@ -58,3 +64,25 @@ func ReadTemplate(path string) (*unstructured.Unstructured, error) {
 	}
 	return obj, nil
 }
+
+// ReadTemplateWithCache creates object template from file specified by the given path
+// or uses cached object if available.
+func ReadTemplateWithCache(path string) (*unstructured.Unstructured, error) {
+	lock.RLock()
+	obj, exists := templateCache[path]
+	lock.RUnlock()
+	if !exists {
+		lock.Lock()
+		defer lock.Unlock()
+		// Recheck codititon.
+		obj, exists = templateCache[path]
```

Comment:
Why you're rechecking condition?

Reply:
Common practice - it is assume that you want to run initialization only once.
It can happen that object is not in cache so thread waits for lock to initialize object. During this waiting time other thread can initialize object. So when first thread acquires lock condition has changed and it's not needed to perform all initialization again. That's why you want to almost always recheck conditions after entering the critical section.

Reply:
Yes, that makes sense.